fix: Single Runtime Chunk and Federation eager module hoisting.

ScriptedAlchemy opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
Hoists Eager Referenced modules into the runtime chunks
<!-- Try to link to an open issue for more information. -->
https://github.com/webpack/webpack/issues/18810
<!-- In addition to that please answer these questions: -->

**What kind of change does this PR introduce?**

<!-- E.g. a bugfix, feature, refactoring, build related change, etc… -->
bugfix
**Did you add tests for your changes?**
no
<!-- Note that we won't merge your changes if you don't add tests -->

**Does this PR introduce a breaking change?**
no
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**What needs to be documented once your changes are merged?**
Nothing
<!-- List all the information that needs to be added to the documentation after merge -->
<!-- When your changes are merged you will be asked to contribute this to the documentation -->

evenstensberg commented:
We will take a look when we have time @ScriptedAlchemy

evenstensberg commented on lib/container/ModuleFederationPlugin.js:
```diff
@@ -34,6 +46,28 @@ class ModuleFederationPlugin {
+			);
+		}
+		let hooks = compilationHooksMap.get(compilation);
+		if (hooks === undefined) {
```
if(!hooks)

ScriptedAlchemy commented:
I just realize i did not add the hoist plugin here

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+	Author Zackary Jackson @ScriptedAlchemy
+*/
+
+"use strict";
```
This plugin needs to be applied still somewhere in the compilation.
https://github.com/module-federation/core/blob/main/packages/enhanced/src/lib/container/runtime/FederationRuntimePlugin.ts like i do here.

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+/**
+ * This class is used to hoist container references in the code.
+ */
+class HoistContainerReferences {
```
in federation v2, we also call hook to add federation v2 runtime as addInclude, hoist container refs also hoists those referenced modules for that dependency as well so that we can ensure runtime plugins etc that user add, plus our own code is also eager regardless of chunk split rules etc. 
https://github.com/module-federation/core/blob/main/packages/enhanced/src/lib/container/ContainerPlugin.ts

Reference hoist ensures all external modules, remote modules, and v2 runtime is all in the runtime chunks and not in come other chunk lower in the tree.

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+		for (const dep of containerEntryDependencies) {
+			const containerEntryModule = moduleGraph.getModule(dep);
+			if (!containerEntryModule) continue;
+			const allReferencedModules = getAllReferencedModules(
```
get all initial referenced modules for each containerEntry dependency or federation runtime dependency, this returns all eager modules not wrapped in import() that the ContainerEntry may need.

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+				"initial"
+			);
+
+			const allRemoteReferences = getAllReferencedModules(
```
Get external modules types like remote modules and externals set in config - if user eager share  or consume module that needs external, it might not have loaded the chunk who had it in the entry so it needs hoisting too.

ScriptedAlchemy replied:
Mostly an issue in single runtime chunk, but if user has library as well in their host code from federation v2, it will occur there

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+			const runtimes = new Set();
+
+			for (const runtimeSpec of containerRuntimes) {
+				compilation.compiler.webpack.util.runtime.forEachRuntime(
```
get the runtimes these modules can belong to, ensure we hoist the module into all runtimes its connected to.

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+
+				for (const module of allReferencedModules) {
+					if (!chunkGraph.isModuleInChunk(module, runtimeChunk)) {
+						chunkGraph.connectChunkAndModule(runtimeChunk, module);
```
connect all the referenced modules to the runtime chunk

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+	cleanUpChunks(compilation, modules) {
+		const { chunkGraph } = compilation;
+		for (const module of modules) {
+			for (const chunk of chunkGraph.getModuleChunks(module)) {
```
remove chunks who might now be empty after module disconnections

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+		const currentModule = stack.pop();
+		if (!currentModule) continue;
+
+		const mgm = compilation.moduleGraph._getModuleGraphModule(currentModule);
```
gives lint / type error but not sure how else to get the modulegraph connections in this way.

ScriptedAlchemy replied:
done

ScriptedAlchemy commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+					const parentBlock = compilation.moduleGraph.getParentBlock(
+						connection.dependency
+					);
+					if (parentBlock instanceof AsyncDependenciesBlock) {
```
skip async imports since chunk handler will be able to get those dependencies ahead of time,  not eager

ScriptedAlchemy commented on lib/container/ContainerPlugin.js:
```diff
@@ -73,6 +74,7 @@ class ContainerPlugin {
 		}
 
 		compiler.hooks.make.tapAsync(PLUGIN_NAME, (compilation, callback) => {
+			const hooks = ModuleFederationPlugin.getCompilationHooks(compilation);
```
These compilation hooks are used in several places in my plugin as well as in my next plugin to handle the modules in various ways. I also have runtime modules that use this as well.

alexander-akait commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+				{
+					name: PLUGIN_NAME,
+					// advanced stage is where SplitChunksPlugin runs.
+					stage: 11 // advanced + 1
```
Let's use `OptimizationStages.STAGE_ADVANCED + 1`

alexander-akait commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,205 @@
+			const runtimes = new Set();
+
+			for (const runtimeSpec of containerRuntimes) {
+				compilation.compiler.webpack.util.runtime.forEachRuntime(
```
We can use `require("utils/runtime')"`, using `compilation.compiler.webpack` required only in external `plugins` (i.e. not in our codebase)

ScriptedAlchemy replied:
perfect

alexander-akait commented:
@ScriptedAlchemy Let's fix test, looks like a typo

Let's also add a couple of tests to show that we've solved the problem

ScriptedAlchemy commented:
Sure. This case is believe this plugin will only solve some aspects. I'll put test and see if it can pass. But ideally I want to get the parts approved first then send a pr that utilizes them. Since there's several files that need to be updated to bring some features back. Initially I wanted review on it. Which is why I don't automatically apply the plugin yet 😅

ScriptedAlchemy commented:
@alexander-akait the tests fail saying `getComplationHooks is not a function` but it is static method on the plugin class. This fails the basictest configCases and crashed the build, can you help me understand why.

<img width="1358" alt="image" src="https://github.com/user-attachments/assets/b5d7ac9a-fba0-4a83-bddf-5d0582a7017e">

alexander-akait commented:
@ScriptedAlchemy Fixed your problem, but it looks like our tests are failed and your test doesn't work either, we need to fix it, I think you can do it faster than me, since you understand the logic better

alexander-akait commented on lib/container/ModuleFederationPlugin.js:
```diff
@@ -87,6 +122,7 @@ class ModuleFederationPlugin {
 					shareScope: options.shareScope
 				}).apply(compiler);
 			}
+			new HoistContainerReferences().apply(compiler);
```
If I understand correctly we do this only for `References`? Perhaps for performance purposes we should add a check:
```
options.remotes &&
				(Array.isArray(options.remotes)
					? options.remotes.length > 0
					: Object.keys(options.remotes).length > 0)
```

like we do for `ContainerReferencePlugin`?

ScriptedAlchemy replied:
I dont think that will work without introducing other problems with eager modules. 

think of this case: 
User shares modules with eager:true, shared modules will only hoist to entrypoint, not runtime - if runtime eagerly share and someone eager require the share - the host runtime will not have the external reference (if the share depends on a external) 

So plugin maybe can be renamed, since its actually HoistEagerExternals. 
In v2, reference hoisting also is responsible for hoisting `@module-federation/runtime,webpack-bundler-runtime,sdk` - since splitChunks will also interfere with federation - this should prevent split chunks from moving eager modules into another chunk beside the runtime.

So while i use it for other things as well, there are cases in federation where it will still fail as anything the system uses or may provide eagerly needs to be in the root chunk at startup, which can be external modules of a shared package as well. 

Id say the only time we dont need to apply it is when there are no seperate `runtimeChunk` added. So like if its just main.js, then webpack already hoists the module to entry chunk. But when there is runtimeChunk, it is only added later in compile, so all modules "hoist" to entry chunk only, not root runtime chunk .

alexander-akait commented on test/configCases/container/reference-hoisting/index.js:
```diff
@@ -0,0 +1,15 @@
+			);
+		});
+	});
+});
```
Can we do more better testing here because this is just a test for the working rather than the actual behaviour?

ScriptedAlchemy replied:
yeah i updated the test, i was blocked on the original issue being unable to import module federation plugin for the hooks

alexander-akait commented on lib/container/HoistContainerReferencesPlugin.js:
```diff
@@ -0,0 +1,211 @@
+	 * @param {import("../Compilation")} compilation The webpack compilation instance.
+	 * @returns {Set<import("../Chunk")>} Set of runtime chunks.
+	 */
+	getRuntimeChunks(compilation) {
```
Also we don't use it in our code

ScriptedAlchemy replied:
done

alexander-akait commented:
To avoid:

> ConfigTestCases › chunk-graph › rewalk-chunk › rewalk-chunk should compile

just make rebase to `main`

ScriptedAlchemy commented:
> @ScriptedAlchemy Fixed your problem, but it looks like our tests are failed and your test doesn't work either, we need to fix it, I think you can do it faster than me, since you understand the logic better

Yeah, no problem. The main blocker that I had was, I could not import the plug in to even start looking at the test. Now that that is resolved, I will actually get the test working and create something that confirms it works

ScriptedAlchemy commented:
@alexander-akait the plugin is now hoisting modules correctly as you can see in the runtime chunk, it is now not empty. 
<img width="1358" alt="image" src="https://github.com/user-attachments/assets/a7bd438d-b00c-40d3-8578-351989032a24">

But i get a test runtime error, why?

<img width="1358" alt="image" src="https://github.com/user-attachments/assets/17aec8a2-abd0-4632-8c46-44c3aa4d359a">

ScriptedAlchemy commented:
Okay i found the problem i have updated the test. 

There is an issue / race condition with how these test work. Since the "host" app consumes its own remoteEntry - when runtiemChunk: single is set in ESM mode, the runtime ends up externally importing ./container.mjs -> import webpack runtime. So it fails in this case, however this pattern is hardly ever used as people are not consuming their own containers,  its usually another build who is. 

I have added a test case to check the modules in the module cache, when there is 2 entrypoints - the external container modules will fail to init that are in the entry who is not loaded and the cache will only include the modules who happen to be loaded at startup, with the hoist plugin enabled the warning goes away and the module cache shows all remote modules regardless of which entry uses them

alexander-akait commented:
@ScriptedAlchemy We still have two problems:
- `chunk-graph › rewalk-chunk › rewalk-chunk should compile` (let's rebase)

This looks like a slight regression, can you look at it and understand why it happened:

- `container › reference-hoisting › exported tests › should have the hoisted container references`
-  container › reference-hoisting › exported tests › should have the hoisted container reference`

ScriptedAlchemy commented:
Okay I'll check

ScriptedAlchemy commented:
@alexander-akait i found the problem, walk chunk test has been corrected. 

It seems somehow, i removed the application of the plugin. I have re-applied it again and now the hoist ref test case passes too

alexander-akait commented:
@ScriptedAlchemy I think we can go further, we should add documentation, but we can do it later, when we finish the work, what do you think about it?

ScriptedAlchemy commented:
> @ScriptedAlchemy I think we can go further, we should add documentation, but we can do it later, when we finish the work, what do you think about it?

Yeah when we finish the work. Then we can document. Some PR have no need for doc or are just mechanical
